tensorflow/python/eager/function.py

```diff
@@ -2525,6 +2535,31 @@ def _convert_variables_to_tensors(self, args, kwargs):
     kwargs = {kw: ops.convert_to_tensor(x) for kw, x in kwargs.items()}
     return tuple(args), kwargs
 
+  def _convert_typed_variables_to_tensors(self, args, kwargs):
+    if self.input_signature is not None:
+      return
+
+    args = list(args)
+    for i, arg in enumerate(args):
+      if i < len(self._fullargspec.args):
+        arg_annotation = self._fullargspec.annotations.get(
+            self._fullargspec.args[i])
+        if arg_annotation == core_tf_types.TensorLike:
+          args[i] = ops.convert_to_tensor(arg)
+      else:
+        varargs_annotation = self._fullargspec.annotations.get(
+            self._fullargspec.varargs)
+        if varargs_annotation == core_tf_types.TensorLike:
+          args[i] = ops.convert_to_tensor(arg)
+
+    if self._fullargspec.varkw is not None:
+      varkw_annotation = self._fullargspec.annotations.get(
+          self._fullargspec.varkw)
+      if varkw_annotation == core_tf_types.TensorLike:
+        kwargs = {kw: ops.convert_to_tensor(x) for kw, x in kwargs.items()}
```

Comment:
Do kwargs always map to varkw? Named arguments should also be placed in kwargs. Can you add a test which only uses named arguments to verify? Something like:

`def f(a: TensorLike = 0, b: int = 1, **kwargs)`

Reply:
Yeah, kwargs always map to varkw. I added tests to verify only named arguments are converted to Tensors.

Reply:
Please add this test:

`def f(x : tf.Tensor = 0, y : int = 1, **kwargs: in.Tensor)`

Called with:

`f(x=1, y=2, z=3)`  # Should trace
`f(x=1, y=3, z=3)`  # Should trace
`f(x=2, y=2, z=4)`  # Should not trace
`f(x=2, y=2, z=4, u=5)`  # Should trace

Reply:
A bit more info here: looking at `kwargs`, you may apply the annotation of `**kwargs`, but not to the arguments specified found in `kwonlyargs`, nor to the argument names found in `args`. This is an example of kwonly args:

`def f(*, a=None)`

And this is an example showing how a positional argument ends up in kwargs:

```
import inspect

def decorator(f):
  print(inspect.getfullargspec(f))
  def wrapper(*args, **kwargs):
    print(args)
    print(kwargs)
  return wrapper

@decorator
def f(a, **kwargs):
  pass

f(a=1)
```

These are a bit tricky to get right, but with sufficient tests we can.

Reply:
Good test case! I mapped the positional arguments found in `kwargs` back to their own annotations. The annotation of `**kwargs` is applied only to arguments found in `kwonlyargs`. Changes are in `265535e`.

I'm going to add more tests to ensure the new logic works for `args` with/without annotations.

Reply:
Test coverage for `args`, `*args`, `kwonlyargs`, and `**kwargs` with/without annotations are in dec9749.